future-apps/payroll/contracts/Payroll.sol

```diff
@@ -1,36 +1,48 @@
-pragma solidity 0.4.18;
+pragma solidity 0.4.24;
 
 import "@aragon/os/contracts/apps/AragonApp.sol";
-import "@aragon/os/contracts/common/Initializable.sol";
-import "@aragon/os/contracts/common/IForwarder.sol";
+// import "@aragon/os/contracts/common/IForwarder.sol";
 
-import "@aragon/os/contracts/lib/zeppelin/token/ERC20.sol";
-import "@aragon/os/contracts/lib/zeppelin/math/SafeMath.sol";
+import "@aragon/os/contracts/lib/token/ERC20.sol";
+import "@aragon/os/contracts/lib/math/SafeMath.sol";
+import "@aragon/os/contracts/lib/math/SafeMath64.sol";
+import "@aragon/os/contracts/lib/math/SafeMath8.sol";
 
-import "@aragon/apps-finance/contracts/Finance.sol";
+import "@aragon/ppf-contracts/contracts/IFeed.sol";
 
-import "./DenominationToken.sol";
+import "@aragon/apps-finance/contracts/Finance.sol";
 
 
 /**
  * @title Payroll in multiple currencies
  */
 contract Payroll is AragonApp { //, IForwarder { // makes coverage crash (removes pure and interface doesnt match)
     using SafeMath for uint256;
-    using DenominationToken for uint256;
+    using SafeMath64 for uint64;
+    using SafeMath8 for uint8;
 
     // kernel roles
     bytes32 constant public ADD_EMPLOYEE_ROLE = keccak256("ADD_EMPLOYEE_ROLE");
-    bytes32 constant public REMOVE_EMPLOYEE_ROLE = keccak256("REMOVE_EMPLOYEE_ROLE");
+    bytes32 constant public TERMINATE_EMPLOYEE_ROLE = keccak256("TERMINATE_EMPLOYEE_ROLE");
+    bytes32 constant public SET_EMPLOYEE_SALARY_ROLE = keccak256("SET_EMPLOYEE_SALARY_ROLE");
+    bytes32 constant public ADD_ACCRUED_VALUE_ROLE = keccak256("ADD_ACCRUED_VALUE_ROLE");
     bytes32 constant public ALLOWED_TOKENS_MANAGER_ROLE = keccak256("ALLOWED_TOKENS_MANAGER_ROLE");
-    bytes32 constant public ORACLE_ROLE = keccak256("ORACLE_ROLE");
+    bytes32 constant public CHANGE_PRICE_FEED_ROLE = keccak256("CHANGE_PRICE_FEED_ROLE");
+    bytes32 constant public MODIFY_RATE_EXPIRY_ROLE = keccak256("MODIFY_RATE_EXPIRY_ROLE");
+
     address constant public ETH = address(0);
+    uint128 constant public ONE = 10 ** 18; // 10^18 is considered 1 in the price feed to allow for decimal calculations
+    uint64 constant public MAX_UINT64 = uint64(-1);
+    uint256 constant public MAX_ACCRUED_VALUE = 2**128;
 
     struct Employee {
         address accountAddress; // unique, but can be changed over time
         mapping(address => uint8) allocation;
-        uint256 denominationTokenSalary; // per second
-        uint256 lastPayroll;
+        uint256 denominationTokenSalary; // per second in denomination Token
+        uint256 accruedValue;
+        uint64 lastPayroll;
+        uint64 endDate;
+        bool terminated;
         string name;
```

Comment:
@sohkai @bingen what do you think about changing name to ens when adding an employee and in the event?
We think it'll be more useful than name, like we could use it for retrieving employee info from identity app

Reply:
@lmcorbalan I think resolving ENS should be a front-end aspect by pulling from the recipient address; the name provided here is really metadata for the employee (I think of it as their "de facto" name in the context of the organization, since an employee can change their recipient address).